### PR TITLE
Known state: store tags on stack as unions.

### PR DESCRIPTION
**Pull request by chriseth**
This helps discovering the return jump address in certain cases and thus retaining state information across returns.

### CODE REVIEW

File: libevmasm/CommonSubexpressionEliminator.cpp [hunk 1]
```diff
@@ -197,7 +199,9 @@ void CSECodeGenerator::addDependencies(Id _c)
 		addDependencies(argument);
 		m_neededBy.insert(make_pair(argument, _c));
 	}
-	if (expr.item->type() == Operation && (
+	if (
+		expr.item &&
+		expr.item->type() == Operation && (
 		expr.item->instruction() == Instruction::SLOAD ||
```
Comment (LianaHus) [inline libevmasm/CommonSubexpressionEliminator.cpp]: intend?

File: libevmasm/KnownState.cpp [hunk 2]
```diff
@@ -212,18 +229,17 @@ ExpressionClasses::Id KnownState::stackElement(int _stackHeight, SourceLocation
 	if (m_stackElements.count(_stackHeight))
 		return m_stackElements.at(_stackHeight);
 	// Stack element not found (not assigned yet), create new unknown equivalence class.
-	//@todo check that we do not infer incorrect equivalences when the stack is cleared partially
-	//in between.
-	return m_stackElements[_stackHeight] = initialStackElement(_stackHeight, _location);
+	return m_stackElements[_stackHeight] =
+			m_expressionClasses->find(AssemblyItem(UndefinedItem, _stackHeight, _location));
```
Comment (LianaHus) [inline libevmasm/KnownState.cpp]: intend

**Comment by chriseth** [comment]
Optimizer test failure on OSX, waiting for confirmation after latest commit.

**Comment by LianaHus** [comment]
looks good but OSX biuld fails

**Comment by chriseth** [comment]
Ran OSX tests locally and they were fine.

**Pull request merged**